### PR TITLE
feat(root): add common registration method

### PR DESCRIPTION
**Pull request by bitgoAaron**
## Description
- makes all coin modules easier to import & register through a common registration pattern
- updates web-demo to demonstrate how to use dynamic import for each coin

## Issue Number

Ticket: BG-64112

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- see web-demo /coins page

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My code compiles correctly for both Node and Browser environments
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [x] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: modules/web-demo/src/components/Coins/coinFactory.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
+import { BitGoAPI } from '@bitgo/sdk-api';
+import { BaseCoin } from '@bitgo/sdk-core';
+
+class CoinFactory {
```
Comment (joey-bai) [inline modules/web-demo/src/components/Coins/coinFactory.ts:4]: should we create a CoinFactory in `sdk-api`?
`sdk-api` already depends on `sdk-core`

**Comment by mmcshinsky-bitgo** [comment]
Am I correct in understanding that you can in essence reverse register coins via the coin package instead of the sdk method for one off coins as well?
```javascript
import { BitGoAPI } from '@bitgo/sdk-api';
import { register } from '@bitgo/sdk-coin-btc';

const sdk = new BitGoAPI()
register(sdk);
```

**Comment by bitgoAaron** [comment]
> Am I correct in understanding that you are reverse registering coins via the coin package instead of the sdk method by referencing sdk variable reference?
> 
> ```js
> import { BitGoAPI } from '@bitgo/sdk-api';
> import { register } from '@bitgo/sdk-coin-btc';
> 
> const sdk = new BitGoAPI()
> register(sdk);
> ```
> 
> For multiple coins I'd have to do something like:
> 
> ```js
> import { register as registerBtc } from '@bitgo/sdk-coin-btc';
> import { register as registerLtc } from '@bitgo/sdk-coin-ltc';
> ```

Correct. It's meant to make dynamic imports easier by ensuring every coin has the same method signature for registration. [See the web-demo factory for how we will implement this in bgms](https://github.com/BitGo/BitGoJS/pull/3118/files#diff-ee9de2a4b053356a8a88521062e3151234323f21180129ea8e3f605a960cf737R30)

File: modules/web-demo/src/components/Coins/coinFactory.ts (hunk 1 above)
Comment (bitgoAaron) [reply modules/web-demo/src/components/Coins/coinFactory.ts:4]: No, `sdk-api` can't import coin specific modules - it must remain decoupled. This class is purely for demonstration purposes and has no bearing on the published code.

File: modules/web-demo/src/components/Coins/coinFactory.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    sdk: BitGoAPI,
+    register: (sdk: BitGoAPI) => void,
+  ) {
+    register(sdk);
+    const coin = sdk.coin(name);
+    this.coinCache.set(name, coin);
+    return coin;
+  }
+
+  async getCoin(name: string, sdk: BitGoAPI): Promise<BaseCoin | undefined> {
```
Comment (lifeiscontent) [inline modules/web-demo/src/components/Coins/coinFactory.ts:22]: shouldn't this always return a BaseCoin? if not, it should  throw an error or something, no?

```suggestion
  async getCoin(name: string, sdk: BitGoAPI): Promise<BaseCoin> {
```

**Review by OttoAllmendinger — APPROVED** [review]

**Review by mmcshinsky-bitgo — APPROVED** [review]

File: modules/sdk-core/src/bitgo/coinFactory.ts [hunk 3]
```diff
@@ -37,9 +37,6 @@ export class CoinFactory {
    * @throws Error
    */
   public register(name: string, coin: CoinConstructor): void {
-    if (this.coinConstructors.has(name)) {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-core/src/bitgo/coinFactory.ts:40]: Do we run the risk of stored run time variable references being lost that may be needed by overwriting the last value in the Map?

Comment (bitgoAaron) [reply modules/sdk-core/src/bitgo/coinFactory.ts:40]: Not sure I follow, but if some client code has reference to an older coin and then re-registers it, the first reference they have would still be valid. It's up to the client to manage these with the modular approach, we just need to make sure we're not throwing unnecessary exceptions anymore.

**Pull request merged**